dts: add SPI bindings for nRF family

Qbicz opened the pull request:
This change adds DTS definition of SPI device for nRF5 chips.

This change also removes SPI pin configuration from Kconfig and moves
it to chip DTS.

Signed-off-by: Filip Kubicz <filip.kubicz@nordicsemi.no>

Qbicz commented:
@anangl, @pdunaj, can you take look?

anangl commented on dts/bindings/spi/nordic,nrf-spi.yaml:
```diff
@@ -0,0 +1,53 @@
+# SPDX-License-Identifier: BSD-5-Clause-Nordic
+#
+---
+title: Nordic nRF5 Family SPI Master node
```
Yes. And this also applies to the commit message.

Qbicz replied:
Fixed.

Qbicz commented on dts/bindings/i2c/nordic,nrf-i2c.yaml:
```diff
@@ -1,11 +1,12 @@
 #
 # Copyright (c) 2018 Aapo Vienamo
+# Copyright (c) 2018 Nordic Semiconductor ASA
```
Yes, this is common in Zephyr files. See e.g. `drivers/flash/soc_flash_nrf.c`

ioannisg replied:
Agree, this is allowed.

Qbicz committed:
dts: update I2C bindings for nRF chips

Change 'nRF5' family names to 'nRF' in I2C DTS.

Signed-off-by: Filip Kubicz <filip.kubicz@nordicsemi.no>

Qbicz commented on dts/arm/nordic/nrf51822.dtsi:
```diff
@@ -74,7 +74,7 @@
 
 		i2c0: i2c@40003000 {
-			compatible = "nordic,nrf5-i2c";
+			compatible = "nordic,nrf-i2c";
```
Done.

Qbicz committed:
dts: add SPI bindings for nRF family

This change adds DTS definition of SPI device for nRF chips.
It also removes SPI pin configuration from Kconfig and moves it to
chip DTS.

Signed-off-by: Filip Kubicz <filip.kubicz@nordicsemi.no>

Qbicz commented on include/dt-bindings/spi/spi.h:
```diff
@@ -0,0 +1,11 @@
+#ifndef __DT_BINDINGS_SPI_SPI_H
+#define __DT_BINDINGS_SPI_SPI_H
+
+#define SPI_BITRATE_STANDARD	1000000	/* 1 Mbit/s */
```
Removed, not needed for now.

Qbicz commented on dts/arm/nordic/nrf51822.dtsi:
```diff
@@ -95,6 +95,34 @@
+		};
+
+		qdec: qdec@40012000 {
+			compatible = "nordic,nrf-qdec";
```
Removed qdec.

Qbicz commented:
@pdunaj @anangl All comments have been addressed. I aligned boards using I2C & SPI to these changes so tests are now passing.

carlescufi reviewed (approved):
LGTM

ioannisg commented on boards/arm/nrf52840_pca10056/nrf52840_pca10056.dts:
```diff
@@ -76,6 +76,34 @@
 	scl-pin = <31>;
 };
 
+&spi0 {
```
Could we do something like in line 47 for UARTE1?

ioannisg replied:
I.e. enable further SPIs based on used Kconfig configuration.

Qbicz replied:
Added. DTS intances of ID higher than 0 are now only added if enabled.

ioannisg replied:
The idea is that we should not need DTC to generate macros for things we do not aim to use. Not generating non-required defines always reduces risk of bugs. The same concept should be applied to .fixup files.

Now, I see that instances with ID 0 could also have conditional definitions (as long as the respective Kconfig options exist). We need to have a consistent approach, I agree.

Qbicz replied:
So what should I do under this pull request? I tried to be consistent with other peripherals configuration. @ioannisg please create a separate issue if you think all DTS instances should be added conditionally.

Qbicz replied:
I feel this conditional approach generates too much overhead in the DTS files. @ioannisg 
 eemember, that that the issue you described exists only for development kits. For application boards, there is usually a single useful hardware config: e.g. sensor1 connected to I2C0, sensor2 to SPI1.

On contrary, for development boards, you have a lot of unused peripherals and their usage change. However, in production the usage of every peripheral is clear. Application boards simply don't have DTS instances that are not needed. What do you think @pdunaj ?

galak replied:
We should be avoiding #ifdef in .dts files as much as possible.  The general rule of thumb for the .dts is for it to enable by default what the board has enabled by default, as well as support for any headers.  @erwango has a writeup or such about what defaults we should be aiming for.

Qbicz replied:
Can you point to such writeup or a set of rules for writing DTS/Kconfig configurations?

pdunaj replied:
Hi, if I had to chose I would avoid adding ifdefs in dts file. If something is define but unused there is little harm in that (file is processed anyway, there will be few unused macros). On the contrary I think we should define everything that board has (is capable of), even if unused.

ioannisg reviewed (changes requested):
Looks good to me; added a question about conditional adding of DTS content

ioannisg commented on boards/arm/nrf52_pca20020/nrf52_pca20020.dts:
```diff
@@ -70,6 +70,7 @@
 	};
 };
 
+#if defined(CONFIG_I2C_1_NRF_TWI) || defined(CONFIG_I2C_1_NRF_TWIM)
```
This -if- is unterminated

Qbicz replied:
Fixed

Qbicz commented:
Tested with `tests/drivers/spi/spi_loopback` for `nrf52_pca10056` and `nrf51_pca10028`.

Qbicz commented:
@ioannisg after discussion with @anangl I removed the commit with conditional adding of DTS instances. These macros generated by DTS are only passed to drivers when a specific instance is enabled via Kconfig. There's no need to wrap DTS instances in Kconfig conditions, as it would create more overhead in code and require programmers to look for defines in driver Kconfigs while writing DTS files.

Qbicz commented:
Resolved conflicts after merge of #9814. @ioannisg, can you update your review?

pdunaj commented:
Hi @ioannisg , can we pull this in?

anangl commented:
And what about the other DK boards: nrf52810_pca10040 an nrf52840_pca10059? Shouldn't some SPI instances be enabled in their .dts files as well to keep consistency?

Qbicz commented on boards/arm/96b_carbon_nrf51/96b_carbon_nrf51_defconfig:
```diff
@@ -29,7 +29,4 @@ CONFIG_SPI_1_NAME="SPI_1"
 CONFIG_SPI_1_IRQ_PRI=1
```
Not needed, fixed.

Qbicz commented:
Added SPI dts instances for nrf52810_pca10040 and nrf52840_pca10059. I don't add all possible instances on dongle nrf52840_pca10059 as it has limited GPIOs available.

Qbicz commented on boards/arm/nrf51_pca10028/nrf51_pca10028.dts:
```diff
@@ -51,6 +51,20 @@
+	miso-pin = <7>;
+};
+
+&spi1 {
```
For development kits I followed the rule to have dts instances for everything that is available.

anangl replied:
You should enable `spi2` in _nrf52_pca10040.dts_ then.

Qbicz replied:
Done.

anangl commented:
@Qbicz 
> I don't add all possible instances on dongle nrf52840_pca10059 as it has limited GPIOs available.

Since this is an exception to the rule you used, maybe it is worth a word of comment in the relevant .dts file?

anangl commented on dts/bindings/spi/nordic,nrf-spi.yaml:
```diff
@@ -0,0 +1,60 @@
+      description: required interrupts
+      generation: define
+
+    sclk-pin:
```
In all our Product Specifications this pin is referred as the "SCK pin". I think we should stick to this convention.

Qbicz commented:
@anangl introduced all the changes you asked for.

anangl commented on tests/drivers/spi/spi_loopback/boards/nrf51_pca10028.conf:
```diff
@@ -1,10 +1,6 @@
 CONFIG_SPI_0_NRF_SPI=y
-CONFIG_SPI_0_NRF_SCK_PIN=4
-CONFIG_SPI_0_NRF_MOSI_PIN=6
-CONFIG_SPI_0_NRF_MISO_PIN=5
```
`CONFIG_SPI_LOOPBACK_CS_CTRL_GPIO_PIN` a few lines below (I can't add the comment there) should be updated, too. Pin 30 is now assigned to [MOSI](https://github.com/zephyrproject-rtos/zephyr/pull/9781/files#diff-dad5e9c5da79781871934105f03f7d92R57).

Qbicz replied:
Fixed, also removed collision in PCA10056.

Qbicz commented on boards/arm/nrf52_pca10040/nrf52_pca10040.dts:
```diff
@@ -53,6 +53,27 @@
+	status = "ok";
+	sck-pin = <11>;
+	mosi-pin = <12>;
+	miso-pin = <13>;
```
Good catch. Removed collisions.

Qbicz commented on boards/arm/reel_board/reel_board.dts:
```diff
@@ -105,6 +105,13 @@
 
+&spi3 {
+	status = "ok";
+	sclk-pin = <19>;
```
Done

Qbicz commented:
@anangl, I fixed the pin assignment collisions.

ioannisg commented on drivers/spi/Kconfig.nrfx:
```diff
@@ -6,6 +6,7 @@
 menuconfig SPI_NRFX
 	bool "nRF SPI nrfx drivers"
 	depends on SOC_FAMILY_NRF
+	select HAS_DTS_SPI
```
This option does not, really, belong here. HAS_DTS_SPI signifies that the target platform supports DTS for SPI configuration. It should be somehow selected by the Board or SOC definition.

I saw that the same mistake - IMHO - is done by similar platforms.

Qbicz replied:
Kconfig and DTS are currently connected in Zephyr. Driver's Kconfig needs to be aware if the DTS supports configuration of some peripheral because otherwise it can provide it's own defaults etc.

If you wish to discuss/modify this approach, let's don't do it in this PR.

ioannisg replied:
>Kconfig and DTS are currently connected in Zephyr. Driver's Kconfig needs to be aware if the DTS supports configuration of some peripheral because otherwise it can provide it's own defaults etc.

I didn't disagree on that; i stressed that DTS support should be provided in the hardware definition instead of the driver selection.

ioannisg replied:
I've acknowledged that this convention spreads out to other SOC/Board vendors

ioannisg replied:
Discussed offline with @anangl. Looks like this convention was introduced as the way to make contributors provide DTS configuration needed by drivers/peripherals, in their platforms (@galak , pls, correct if I am wrong).
